// @soroban-react/freighter/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.freighter = freighter;
const freighter_api_1 = __importDefault(require("@stellar/freighter-api"));
/**
 * Returns a connector object for the Freighter wallet.
 * @returns {Connector} - The connector object.
 */
function freighter() {
    return {
        id: 'freighter',
        name: 'Freighter',
        iconUrl: 'https://stellar.creit.tech/wallet-icons/freighter.svg',
        // iconUrl: async () => (await import('./freighter.svg')).default,
        iconBackground: '#fff',
        // TODO: Check this
        installed: true,
        downloadUrls: {
            browserExtension: 'https://chrome.google.com/webstore/detail/freighter/bcacfldlkkdogcmkkibnjlakofdplcbk?hl=en',
        },
        isConnected() {
            return freighter_api_1.default === null || freighter_api_1.default === void 0 ? void 0 : freighter_api_1.default.isConnected();
        },
        getNetworkDetails() {
            return freighter_api_1.default.getNetworkDetails();
        },
        getPublicKey() {
            return freighter_api_1.default.getPublicKey();
        },
        signTransaction(xdr, opts) {
            return freighter_api_1.default.signTransaction(xdr, opts);
        },
    };
}
